diff --git a/src/aes_hardware.c b/src/aes_hardware.c
index e98feb1..ce39de4 100644
--- a/src/aes_hardware.c
+++ b/src/aes_hardware.c
@@ -22,7 +22,11 @@
 /* --- CLang --- */
 #elif defined(__clang__)
 
-#if __has_attribute(target) && __has_include(<wmmintrin.h>) && (defined(__x86_64__) || defined(__i386))
+#if defined(__APPLE__) && defined(TARGET_OS_SIMULATOR) && TARGET_OS_SIMULATOR
+#define IS_APPLE_SIMULATOR
+#endif
+
+#if __has_attribute(target) && __has_include(<wmmintrin.h>) && (defined(__x86_64__) || defined(__i386)) && !defined(IS_APPLE_SIMULATOR)
 #define HAS_AES_HARDWARE AES_HARDWARE_NI
 
 #elif __has_attribute(target) && __has_include(<arm_neon.h>) && (defined(__aarch64__))
diff --git a/src/chacha20poly1305.c b/src/chacha20poly1305.c
index b1fdcd3..b51ba41 100644
--- a/src/chacha20poly1305.c
+++ b/src/chacha20poly1305.c
@@ -369,7 +369,7 @@ fail:
   return 0;
 }
 
-#if defined(__APPLE__) && defined(__MAC_10_12)
+#if defined(__APPLE__) && defined(__MAC_10_12) && __MAC_OS_X_VERSION_MAX_ALLOWED >= __MAC_10_12
 #include <sys/random.h>
 #endif
 
